### DISCUSSION TITLE
Which generated type i can use for prisma model entities including all their relation types?

### DISCUSSION THREAD

**Discussion by andreasvh-conceto** [prisma/prisma#18216] [Q&A] [answered]
Hello, 

assuming i have the following prisma models:
```prisma
model AppUser {
  id        BigInt @id @default(autoincrement())
  firstname String
  lastname  String
  email     String
  posts     Post[]

  @@map("app_user")
}

model Post {
  id             BigInt       @id @default(autoincrement())
  comment        String
  userId         BigInt
  user           AppUser      @relation(fields: [userId], references: [id])
  postArtifactId BigInt       @map("post_artifact_id")
  postArtifact   PostArtifact @relation(fields: [postArtifactId], references: [id])

  @@map("post")
}

model PostArtifact {
  id    BigInt @id @default(autoincrement())
  name  String
  posts Post[]

  @@map("post_artifact")
}
```

After performing npx prisma generate on the new prisma version, i saw, that in newer prisma versions the generated types changed a little bit to:
```ts
export type AppUserPayload<ExtArgs extends runtime.Types.Extensions.Args = runtime.Types.Extensions.DefaultArgs> = {
  objects: {
    posts: PostPayload<ExtArgs>[]
  }
  scalars: runtime.Types.Extensions.GetResult<{
    id: bigint
    firstname: string
    lastname: string
    email: string
  }, ExtArgs['result']['appUser']>
}

/**
 * Model AppUser
 * 
 */
export type AppUser = AppUserPayload['scalars']
```
In earlier prisma versions only the basic attributes (without references were defined here). Now i also see that the referenced objects are also defined in this type AppUserPayload['objects'].

I want to create an own interface for the appUser model with all nested relations (and nested nested relations) it has. I currently did that in a way like this:

```ts
export interface IAppUser extends AppUser {
  posts: IPost[] | null;
}

export interface IPost extends Post {
  postArtifact: IPostArtifact | null;
}

export interface IPostArtifact extends PostArtifact {
  post: IPost | null;
}
```

or with another approach like this:

```ts
// DO NOT use this method, it is just there to receive a correct prisma type
async function getFullUserEntityPromise() {
  return new PrismaClient().appUser.findFirstOrThrow({
    include: {
      posts: {
        include: {
          postArtifact: true,
        },
      },
    },
  });
}

export type UserWithReferencePromise = ReturnType<
  typeof getFullUserEntityPromise
>;

export type UserWithReferences = UserWithReferencePromise extends Promise<
  infer T
>
  ? T
  : never;
```

For us it is important to have concrete interfaces in order to reference them in method return types, in generic structures like maps, sets etc.. Since i saw that the generated type has changed and holds the AppUserPayload['objects'], is there now a cleaner way, or better, a out of the box type i can use for getting an Prisma model interface with all its relations it has? 

Or is there something planned to export such kind of type? It would be really helpful, because defining so many interfaces all the time can be annoying for each model.

Thanks a lot for your help and best regards

Andreas

**Comment by nurul3101** [comment, marked as answer]
Hello @andreasvh-conceto 👋 

This type is generated because of Client Extensions, It looks like you have Client extensions enabled, If you would remove the Client Extensions then the types with only scalar fields would be generated.
```ts
export type AppUserPayload<ExtArgs extends runtime.Types.Extensions.Args = runtime.Types.Extensions.DefaultArgs> = {
  objects: {
    posts: PostPayload<ExtArgs>[]
  }
  scalars: runtime.Types.Extensions.GetResult<{
    id: bigint
    firstname: string
    lastname: string
    email: string
  }, ExtArgs['result']['appUser']>
}
```

Would creating a type like this work for your use case? ([Reference](https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/operating-against-partial-structures-of-model-types#problem-using-variations-of-the-generated-model-type))
```ts
const appUserWithPostsAndArtifacts = Prisma.validator<Prisma.AppUserArgs>()({
  include: {
    posts: {
      include: {
        postArtifact: true,
      },
    },
  },
});
```

**Reply by nurul3101** [reply]
@andreasvh-conceto If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by leosavvy** [reply]
@nurul3101 From what I could see on the web the only way of enabling client extensions is through the schema via feature preview and by using the $extends method. I could not find either of these on my code and still getting this error, can you redirect me to a solution for disabling Client Extensions on the entire project? 

Thanks!

**Reply by nurul3101** [reply]
@leosavvy Which version of prisma are you using? Can you share the error message that you are getting?
On a side note, could you open a new [GitHub Discussion](https://github.com/prisma/prisma/discussions/new), so that we can move the conversation there?

Thanks!

**Reply by leosavvy** [reply]
@nurul3101 I've just created the discussion as you requested: https://github.com/prisma/prisma/discussions/20009

Thank you!

**Reply by nurul3101** [reply]
Thanks 🙌 
I have replied to the discussion, let's continue conversation in that thread 🙂
